File: app/src/main/java/com/jimzrt/umsmounter/Fragments/MainFragment.java
package com.jimzrt.umsmounter.Fragments;

import android.app.AlertDialog;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SimpleItemAnimator;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageButton;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.jimzrt.umsmounter.Activities.MainActivity;
import com.jimzrt.umsmounter.ListAdapters.ImageListAdapter;
import com.jimzrt.umsmounter.Model.ImageItem;
import com.jimzrt.umsmounter.R;
import com.jimzrt.umsmounter.Tasks.BaseTask;
import com.jimzrt.umsmounter.Tasks.MountImageTask;
import com.jimzrt.umsmounter.Tasks.UnmountingTask;
import com.jimzrt.umsmounter.Utils.BackgroundTask;
import com.jimzrt.umsmounter.Utils.Helper;
import com.jimzrt.umsmounter.ViewModels.ImageItemViewModel;
import com.tonyodev.fetch2.Download;
import com.tonyodev.fetch2.Fetch;
import com.tonyodev.fetch2.FetchListener;
import com.tonyodev.fetch2.NetworkType;
import com.tonyodev.fetch2.Priority;
import com.tonyodev.fetch2.Request;
import com.topjohnwu.superuser.Shell;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * A placeholder fragment containing a simple view.
 */
public class MainFragment extends Fragment implements ImageListAdapter.OnImageListListener {



    // Animation animation;
    RecyclerView listView = null;
    ImageListAdapter listViewAdapter = null;
    Spinner usbMode = null;
    ArrayAdapter<String> usbModeAdapter = null;
    private boolean populate;
    private String functionMode = "mtp,adb";
    private Fetch mainFetch;
    private ImageItemViewModel model;


    @Override
    public void onActivityCreated(Bundle bundle) {
        super.onActivityCreated(bundle);
        if (populate) {
            populate = false;
        }
    }


    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        if (!hidden) {
            if (getActivity() != null) {
                ((AppCompatActivity) getActivity()).getSupportActionBar().setTitle("UMS Mounter");
            }
        }

    }

    @Override
    public void onResume() {
        super.onResume();


    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mainFetch.close();
    }


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {


        if (mainFetch == null) {
            mainFetch = new Fetch.Builder(getContext(), "Main")
                    .setDownloadConcurrentLimit(4) // Allows Fetch to download 4 downloads in Parallel.
                    .enableLogging(true)
                    .build();
            mainFetch.cancelAll();
            mainFetch.removeAll();

        }

        View view = inflater.inflate(R.layout.fragment_main, container, false);


        if (getActivity() != null)
            ((AppCompatActivity) getActivity()).getSupportActionBar().setTitle("UMS Mounter");
        listView = view.findViewById(R.id.listview);
        LinearLayoutManager mLayoutManager = new LinearLayoutManager(getActivity());
        listView.setLayoutManager(mLayoutManager);
        model = ViewModelProviders.of(this).get(ImageItemViewModel.class);

        listViewAdapter = new ImageListAdapter(model.getImages(false).getValue(), getContext(), this);


        ((SimpleItemAnimator) listView.getItemAnimator()).setSupportsChangeAnimations(false);
        listView.setAdapter(listViewAdapter);
        listView.setHasFixedSize(true);
        listView.getItemAnimator().setChangeDuration(0);
        listView.getItemAnimator().setAddDuration(500);
        listView.getItemAnimator().setMoveDuration(500);
        listView.getItemAnimator().setRemoveDuration(500);
        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(listView.getContext(),
                mLayoutManager.getOrientation());
        listView.addItemDecoration(dividerItemDecoration);


        usbMode = view.findViewById(R.id.spinner);

        usbModeAdapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_dropdown_item,
                new ArrayList<>());


        usbMode.setAdapter(usbModeAdapter);


        ImageButton refreshButton = view.findViewById(R.id.buttonRefresh);

        refreshButton.setOnClickListener((View v) -> {
            // populateList();
            listViewAdapter.addItems(model.getImages(true).getValue());
            Toast.makeText(getContext(), "List updated!", Toast.LENGTH_LONG).show();
        });


        SwipeRefreshLayout mySwipeRefreshLayout = view.findViewById(R.id.swiperefresh);

        mySwipeRefreshLayout.setOnRefreshListener(
                () -> {

                    listViewAdapter.addItems(model.getImages(true).getValue());
                    mySwipeRefreshLayout.setRefreshing(false);
                    Toast.makeText(getContext(), "List updated!", Toast.LENGTH_LONG).show();
                }
        );


        SharedPreferences sharedPref = getContext().getSharedPreferences(null, Context.MODE_PRIVATE);
        TextView textMode = view.findViewById(R.id.textMode);
        textMode.setText(sharedPref.getString("usbMode", "Not Supported"));


        List<String> modes = new ArrayList<>(Arrays.asList("Writable USB", "Read-only USB"));
        if (sharedPref.getBoolean("cdrom", false)) {
            modes.add("CD-ROM");
        }

        usbModeAdapter.clear();
        usbModeAdapter.addAll(modes);
        usbMode.setSelection(0);

        TextView statusText = view.findViewById(R.id.statusText);
        statusText.setText("Nothing mounted");
        statusText.setTextColor(Color.LTGRAY);


        model.getSelected().observe(this, image -> {
            listViewAdapter.setSelectedItem(image);
        });
        model.getDownloading().observe(this, image -> {
            Log.i("lala", "position: " + listViewAdapter.getPositionOfItem(image));
            listViewAdapter.notifyItemChanged(listViewAdapter.getPositionOfItem(image), "download");
        });
        model.getRemoved().observe(this, image -> {
            Log.i("lala", "REMOVEDDD");
            File file = new File(image.getUserPath());
            file.delete();
            listViewAdapter.remove(image);
        });
        model.getAdded().observe(this, image -> {
            Log.i("lala", "ADDEDDD");
            listView.smoothScrollToPosition(listViewAdapter.addItem(image));
        });
        model.getMounted().observe(this, image -> {
            if (image.getMounted()) {
                Log.i("lala", "Mounted");
                statusText.setText(image.getName() + " mounted");
                statusText.setTextColor(Color.DKGRAY);
                listViewAdapter.notifyItemChanged(listViewAdapter.getPositionOfItem(image), "mount");
            } else {
                Log.i("lala", "Unounted");
                statusText.setText("Nothing mounted");
                statusText.setTextColor(Color.LTGRAY);

                listViewAdapter.notifyItemChanged(listViewAdapter.getPositionOfItem(image), "unmount");
            }

        });


        return view;
    }

    public void unmount(String functionMode) {
        (new BackgroundTask(getActivity())).setDelegate((successful, result) -> {
            if (successful) {
                model.unmount();

            } else {
                AlertDialog.Builder builder = new AlertDialog.Builder(getContext());

                builder.setMessage(result)
                        .setTitle("Error!");
                builder.setPositiveButton("Ok", null);
                AlertDialog dialog = builder.create();
                dialog.show();
            }
        }).setTasks(new BaseTask[]{new UnmountingTask(functionMode)}).execute();
    }

    public void mount(ImageItem imageItem) {


        List<String> output = Shell.Sync.sh("getprop sys.usb.config");
        functionMode = output.get(0);
        if (functionMode == null || functionMode.contains("mass_storage")) {
            functionMode = "mtp,adb";
        }
        (new BackgroundTask(getActivity())).setDelegate((successful, result) -> {
            if (successful) {
                model.mount(imageItem);

            } else {
                AlertDialog.Builder builder = new AlertDialog.Builder(getContext());

                builder.setMessage(result)
                        .setTitle("Error!");
                builder.setPositiveButton("Ok", null);
                AlertDialog dialog = builder.create();
                dialog.show();
            }
        }).setTasks(new BaseTask[]{new MountImageTask(imageItem, usbMode.getSelectedItem().toString())}).execute();

    }


    @Override
    public void onMountImageButtonClicked() {

        ImageItem mounted = model.getMounted().getValue();
        ImageItem selected = model.getSelected().getValue();
        if (mounted == null || !mounted.getMounted()) {
            mount(selected);
        } else if (mounted != selected) {
            unmount(functionMode);
            mount(selected);
        } else {
            unmount(functionMode);
        }


    }

    @Override
    public void onDeleteImageButtonClicked() {

        ImageItem checkedItem = model.getSelected().getValue();

        AlertDialog.Builder adb = new AlertDialog.Builder(getActivity());


        adb.setTitle("Delete");


        adb.setIcon(android.R.drawable.ic_dialog_alert);
        adb.setMessage("Do you really want to delete " + checkedItem.getName() + "?");


        adb.setPositiveButton("Delete", (dialog, which) -> {
            if (checkedItem.getDownloadId() != -1) {
                mainFetch.cancel(checkedItem.getDownloadId());
                mainFetch.remove(checkedItem.getDownloadId());
            }

            Toast.makeText(getContext(), "Deleted!", Toast.LENGTH_SHORT).show();
            model.unselect();
            model.remove(checkedItem);
        });


        adb.setNegativeButton("Cancel", null);
        adb.show();


    }

    @Override
    public void onImageSelected(ImageItem item) {
        model.select(item);
    }

    public void createImage(ImageItem destFile) {

        model.unselect();
        model.add(destFile);


        destFile.setDownloading(true);


        final int[] oldProgress = {0};
        Thread createImageThread = new Thread(() -> {
            try {

                String sourceFile = getContext().getCacheDir().getAbsolutePath() + "/fat.img";
                FileInputStream fis = null;
                FileOutputStream fos = null;
                long size = 0;

                try {
                    fis = new FileInputStream(sourceFile);
                    fos = new FileOutputStream(destFile.getUserPath(), true);

                    size = fis.getChannel().size() + fos.getChannel().size();
                    byte[] buffer = new byte[1024 * 512];
                    long pos = (int) Math.ceil(fos.getChannel().size() / buffer.length);

                    int noOfBytes = 0;


                    // read bytes from source file and write to destination file
                    while ((noOfBytes = fis.read(buffer)) != -1) {
                        // if(pos % 4 == 0){

                        pos += 1;

                        long finalPos = pos;
                        int progress = (int) ((finalPos * buffer.length * 100L) / size);
                        destFile.setProgress(progress);
                        destFile.setSize(Helper.humanReadableByteCount((finalPos * buffer.length)));
                        if ((progress - oldProgress[0]) > 1) {
                            oldProgress[0] = progress;

                            model.downloading(destFile);


                        }


                        fos.write(buffer, 0, noOfBytes);
                    }
                } catch (FileNotFoundException e) {
                    System.out.println("File not found" + e);
                } catch (IOException ioe) {
                    System.out.println("Exception while copying file " + ioe);
                } finally {
                    // close the streams using close method
                    try {
                        if (fis != null) {
                            fis.close();
                        }
                        if (fos != null) {
                            fos.close();
                        }
                    } catch (IOException ioe) {
                        System.out.println("Error while closing stream: " + ioe);
                    }

                    File src = new File(sourceFile);
                    src.delete();


                }


                //
                long finalSize = size;
                getActivity().runOnUiThread(() -> {
                    Toast.makeText(getActivity(), "Image successfully created", Toast.LENGTH_LONG).show();
                    destFile.setDownloading(false);
                    destFile.setSize(Helper.humanReadableByteCount(finalSize));
                    model.downloading(destFile);
                    listView.smoothScrollToPosition(listViewAdapter.getPositionOfItem(destFile));

                    model.select(destFile);

                });

            } catch (Exception e) {
                e.printStackTrace();
            }
        });
        createImageThread.setDaemon(true);
        createImageThread.start();


    }


    public void addImage(ImageItem imageItem) {


        model.unselect();
        model.remove(imageItem);
        model.add(imageItem);

        imageItem.setDownloading(true);


        final Request request = new Request(imageItem.getUrl(), MainActivity.USERPATH + "/" + imageItem.getName());
        request.setPriority(Priority.HIGH);
        request.setNetworkType(NetworkType.WIFI_ONLY);

        imageItem.setDownloadId(request.getId());


        final FetchListener fetchListener = new FetchListener() {
            int oldProgress = 0;

            @Override
            public void onQueued(@NotNull Download download) {
                if (request.getId() == download.getId()) {


                }
            }

            @Override
            public void onCompleted(@NotNull Download download) {
                imageItem.setDownloading(false);
                imageItem.setSize(Helper.humanReadableByteCount(download.getTotal()));
                model.downloading(imageItem);
                listView.smoothScrollToPosition(listViewAdapter.getPositionOfItem(imageItem));
                model.select(imageItem);

                getActivity().runOnUiThread(() -> {
                    Toast.makeText(getContext(), "Downloaded!", Toast.LENGTH_SHORT).show();
                });
                mainFetch.removeListener(this);
                mainFetch.remove(download.getId());
            }

            @Override
            public void onError(@NotNull Download download) {
                model.remove(imageItem);

                getActivity().runOnUiThread(() -> {
                    File file = new File(imageItem.getUserPath());
                    file.delete();
                    Toast.makeText(getContext(), "Deleted!", Toast.LENGTH_SHORT).show();
                    mainFetch.removeListener(this);
                    mainFetch.remove(download.getId());
                });

            }

            @Override
            public void onProgress(@NotNull Download download, long etaInMilliSeconds, long downloadedBytesPerSecond) {
                if (request.getId() == download.getId()) {


                    oldProgress = download.getProgress();

                    imageItem.setProgress(download.getProgress());
                    imageItem.setSize(Helper.humanReadableByteCount(download.getDownloaded()) + " / " + Helper.humanReadableByteCount(download.getTotal()) + " - " + Helper.humanReadableByteCount(downloadedBytesPerSecond) + "/s");


                    model.downloading(imageItem);


                    final int progress = download.getProgress();
                    Log.d("Fetch", "Progress Completed :" + progress);
                }

            }

            @Override
            public void onPaused(@NotNull Download download) {

            }

            @Override
            public void onResumed(@NotNull Download download) {

            }

            @Override
            public void onCancelled(@NotNull Download download) {

            }

            @Override
            public void onRemoved(@NotNull Download download) {

            }

            @Override
            public void onDeleted(@NotNull Download download) {

            }
        };

        mainFetch.addListener(fetchListener);
        mainFetch.enqueue(request, download -> Log.i("lala", "added " + download.getId()), error -> Log.i("lala", "error  " + error.name()));

    }
}

File: app/src/main/java/com/jimzrt/umsmounter/fragments/MainFragment.kt
package com.jimzrt.umsmounter.fragments

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SimpleItemAnimator
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.jimzrt.umsmounter.R
import com.jimzrt.umsmounter.activities.MainActivity
import com.jimzrt.umsmounter.listadapters.ImageListAdapter
import com.jimzrt.umsmounter.listadapters.ImageListAdapter.OnImageListListener
import com.jimzrt.umsmounter.model.ImageItem
import com.jimzrt.umsmounter.tasks.MountImageTask
import com.jimzrt.umsmounter.tasks.UnmountingTask
import com.jimzrt.umsmounter.utils.BackgroundTask
import com.jimzrt.umsmounter.utils.Helper
import com.jimzrt.umsmounter.viewmodels.ImageItemViewModel
import com.tonyodev.fetch2.*
import com.tonyodev.fetch2.Fetch.Impl.getInstance
import com.tonyodev.fetch2core.DownloadBlock
import com.tonyodev.fetch2core.Func
import com.topjohnwu.superuser.Shell
import java.io.*
import java.util.*
import kotlin.math.ceil

/**
 * A placeholder fragment containing a simple view.
 */
class MainFragment : Fragment(), OnImageListListener {
    // Animation animation;
    private var listView: RecyclerView? = null
    private var listViewAdapter: ImageListAdapter? = null
    private var usbMode: Spinner? = null
    private var populate = false
    private var functionMode: String? = "mtp,adb"
    private var mainFetch: Fetch? = null
    private var model: ImageItemViewModel? = null
    override fun onActivityCreated(bundle: Bundle?) {
        super.onActivityCreated(bundle)
        if (populate) {
            populate = false
        }
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (!hidden) {
            if (activity != null) {
                (activity as AppCompatActivity?)!!.supportActionBar!!.title = "UMS Mounter"
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        mainFetch!!.close()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        if (mainFetch == null) {
            val fetchConfiguration = FetchConfiguration.Builder(context!!)
                    .setDownloadConcurrentLimit(4) // Allows Fetch to download 4 downloads in Parallel.
                    .enableLogging(true)
                    .build()
            mainFetch = getInstance(fetchConfiguration)
            mainFetch!!.cancelAll()
            mainFetch!!.removeAll()
        }
        val view = inflater.inflate(R.layout.fragment_main, container, false)
        if (activity != null) (activity as AppCompatActivity?)!!.supportActionBar!!.title = "UMS Mounter"
        listView = view.findViewById(R.id.listview)
        val mLayoutManager = LinearLayoutManager(activity)
        listView?.layoutManager = mLayoutManager
        model = ViewModelProvider(this).get(ImageItemViewModel::class.java)
        listViewAdapter = ImageListAdapter(model!!.getImages(false).value, context!!, this)
        (listView?.itemAnimator as SimpleItemAnimator?)!!.supportsChangeAnimations = false
        listView?.adapter = listViewAdapter
        listView?.setHasFixedSize(true)
        listView?.itemAnimator!!.changeDuration = 0
        listView?.itemAnimator!!.addDuration = 500
        listView?.itemAnimator!!.moveDuration = 500
        listView?.itemAnimator!!.removeDuration = 500
        val dividerItemDecoration = DividerItemDecoration(listView!!.context,
                mLayoutManager.orientation)
        listView?.addItemDecoration(dividerItemDecoration)
        usbMode = view.findViewById(R.id.spinner)
        val usbModeAdapter = ArrayAdapter(context!!, android.R.layout.simple_spinner_dropdown_item,
                ArrayList<String>())
        usbMode?.adapter = usbModeAdapter
        val refreshButton = view.findViewById<ImageButton>(R.id.buttonRefresh)
        refreshButton.setOnClickListener {
            // populateList();
            listViewAdapter!!.addItems(model!!.getImages(true).value!!)
            Toast.makeText(context, "List updated!", Toast.LENGTH_LONG).show()
        }
        val mySwipeRefreshLayout: SwipeRefreshLayout = view.findViewById(R.id.swiperefresh)
        mySwipeRefreshLayout.setOnRefreshListener {
            listViewAdapter!!.addItems(model!!.getImages(true).value!!)
            mySwipeRefreshLayout.isRefreshing = false
            Toast.makeText(context, "List updated!", Toast.LENGTH_LONG).show()
        }
        val sharedPref = context!!.getSharedPreferences(null, Context.MODE_PRIVATE)
        val textMode = view.findViewById<TextView>(R.id.textMode)
        textMode.text = sharedPref.getString("usbMode", "Not Supported")
        val modes: MutableList<String> = ArrayList(Arrays.asList("Writable USB", "Read-only USB"))
        if (sharedPref.getBoolean("cdrom", false)) {
            modes.add("CD-ROM")
        }
        usbModeAdapter.clear()
        usbModeAdapter.addAll(modes)
        usbMode?.setSelection(0)
        val statusText = view.findViewById<TextView>(R.id.statusText)
        statusText.text = "Nothing mounted"
        statusText.setTextColor(Color.LTGRAY)
        model!!.selected.observe(viewLifecycleOwner, Observer { image: ImageItem? -> listViewAdapter!!.setSelectedItem(image) })
        model!!.downloading.observe(viewLifecycleOwner, Observer { image: ImageItem? ->
            Log.i("lala", "position: " + listViewAdapter!!.getPositionOfItem(image!!))
            listViewAdapter!!.notifyItemChanged(listViewAdapter!!.getPositionOfItem(image), "download")
        })
        model!!.removed.observe(viewLifecycleOwner, Observer { image: ImageItem ->
            Log.i("lala", "REMOVEDDD")
            val file = File(image.userPath)
            file.delete()
            listViewAdapter!!.remove(image)
        })
        model!!.added.observe(viewLifecycleOwner, Observer { image: ImageItem? ->
            Log.i("lala", "ADDEDDD")
            listView?.smoothScrollToPosition(listViewAdapter!!.addItem(image!!))
        })
        model!!.mounted.observe(viewLifecycleOwner, Observer { image: ImageItem? ->
            if (image?.mounted!!) {
                Log.i("lala", "Mounted")
                statusText.text = image.name + " mounted"
                statusText.setTextColor(Color.DKGRAY)
                listViewAdapter!!.notifyItemChanged(listViewAdapter!!.getPositionOfItem(image), "mount")
            } else {
                Log.i("lala", "Unounted")
                statusText.text = "Nothing mounted"
                statusText.setTextColor(Color.LTGRAY)
                listViewAdapter!!.notifyItemChanged(listViewAdapter!!.getPositionOfItem(image), "unmount")
            }
        })
        return view
    }

    fun unmount(functionMode: String?) {

        activity?.let {
            BackgroundTask(it).setDelegate(object : BackgroundTask.AsyncResponse {
                override fun processFinish(successful: Boolean?, output: String?) {
                    if (successful!!) {
                        model!!.unmount()
                    } else {
                        val builder = AlertDialog.Builder(context)
                        builder.setMessage(output)
                                .setTitle("Error!")
                        builder.setPositiveButton("Ok", null)
                        val dialog = builder.create()
                        dialog.show()
                    }
                }

            }).setTasks(arrayOf(UnmountingTask(functionMode!!))).execute()
        }
    }

    private fun mount(imageItem: ImageItem?) {
        val output = Shell.su("getprop sys.usb.config").exec().out
        functionMode = output[0]
        if (functionMode == null || functionMode!!.contains("mass_storage")) {
            functionMode = "mtp,adb"
        }
        activity?.let {
            BackgroundTask(it).setDelegate(object : BackgroundTask.AsyncResponse {
                override fun processFinish(successful: Boolean?, output: String?) {
                    if (successful!!) {
                        if (imageItem != null) {
                            model!!.mount(imageItem)
                        }
                    } else {
                        val builder = AlertDialog.Builder(context)
                        builder.setMessage(output)
                                .setTitle("Error!")
                        builder.setPositiveButton("Ok", null)
                        val dialog = builder.create()
                        dialog.show()
                        if (imageItem != null) {
                            model!!.unmount(imageItem)
                        }
                    }
                }
            }).setTasks(arrayOf(MountImageTask(imageItem!!, usbMode!!.selectedItem.toString()))).execute()
        }
    }

    override fun onMountImageButtonClicked() {
        val mounted = model!!.mounted.value
        val selected = model!!.selected.value
        if (mounted == null || !mounted.mounted) {
            mount(selected)
        } else if (mounted !== selected) {
            unmount(functionMode)
            mount(selected)
        } else {
            unmount(functionMode)
        }
    }

    override fun onDeleteImageButtonClicked() {
        val checkedItem = model!!.selected.value
        val adb = AlertDialog.Builder(activity)
        adb.setTitle("Delete")
        adb.setIcon(android.R.drawable.ic_dialog_alert)
        adb.setMessage("Do you really want to delete " + checkedItem!!.name + "?")
        adb.setPositiveButton("Delete") { _: DialogInterface?, _: Int ->
            if (checkedItem.downloadId != -1) {
                mainFetch!!.cancel(checkedItem.downloadId)
                mainFetch!!.remove(checkedItem.downloadId)
            }
            Toast.makeText(context, "Deleted!", Toast.LENGTH_SHORT).show()
            model!!.unselect()
            model!!.remove(checkedItem)
        }
        adb.setNegativeButton("Cancel", null)
        adb.show()
    }

    override fun onImageSelected(item: ImageItem?) {
        model!!.select(item)
    }

    fun createImage(destFile: ImageItem) {
        model!!.unselect()
        model!!.add(destFile)
        destFile.isDownloading = true
        val oldProgress = intArrayOf(0)
        val createImageThread = Thread(Runnable {
            try {
                val sourceFile = MainActivity.USERPATH + "/fat.img"
                var fis: FileInputStream? = null
                var fos: FileOutputStream? = null
                var size: Long = 0
                try {
                    fis = FileInputStream(sourceFile)
                    fos = FileOutputStream(destFile.userPath, true)
                    size = fis.channel.size() + fos.channel.size()
                    val buffer = ByteArray(1024 * 512)
                    var pos: Long = ceil(fos.channel.size() / buffer.size.toDouble()).toLong()
                    var noOfBytes: Int


                    // read bytes from source file and write to destination file
                    while (fis.read(buffer).also { noOfBytes = it } != -1) {
                        // if(pos % 4 == 0){
                        pos += 1
                        val finalPos = pos
                        val progress = (finalPos * buffer.size * 100L / size).toInt()
                        destFile.progress = progress
                        destFile.size = Helper.humanReadableByteCount(finalPos * buffer.size)
                        if (progress - oldProgress[0] > 1) {
                            oldProgress[0] = progress
                            model!!.downloading(destFile)
                        }
                        fos.write(buffer, 0, noOfBytes)
                    }
                } catch (e: FileNotFoundException) {
                    println("File not found$e")
                } catch (ioe: IOException) {
                    println("Exception while copying file $ioe")
                } finally {
                    // close the streams using close method
                    try {
                        fis?.close()
                        fos?.close()
                    } catch (ioe: IOException) {
                        println("Error while closing stream: $ioe")
                    }
                    val src = File(sourceFile)
                    src.delete()
                }


                //
                val finalSize = size
                activity!!.runOnUiThread {
                    Toast.makeText(activity, "Image successfully created", Toast.LENGTH_LONG).show()
                    destFile.isDownloading = false
                    destFile.size = Helper.humanReadableByteCount(finalSize)
                    model!!.downloading(destFile)
                    listView!!.smoothScrollToPosition(listViewAdapter!!.getPositionOfItem(destFile))
                    model!!.select(destFile)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        })
        createImageThread.isDaemon = true
        createImageThread.start()
    }

    fun addImage(imageItem: ImageItem) {
        model!!.unselect()
        model!!.remove(imageItem)
        model!!.add(imageItem)
        imageItem.isDownloading = true
        val request = Request(imageItem.url!!, MainActivity.USERPATH + "/" + imageItem.name)
        request.priority = Priority.HIGH
        request.networkType = NetworkType.WIFI_ONLY
        imageItem.downloadId = request.id
        val fetchListener: FetchListener = object : FetchListener {
            override fun onWaitingNetwork(download: Download) {
                activity!!.runOnUiThread { Toast.makeText(context, "Waiting for Network!", Toast.LENGTH_SHORT).show() }
            }

            override fun onStarted(download: Download, downloadBlocks: List<DownloadBlock>, totalBlocks: Int) {
                activity!!.runOnUiThread { Toast.makeText(context, "Download started!", Toast.LENGTH_SHORT).show() }
            }

            override fun onError(download: Download, error: Error, throwable: Throwable?) {
                model!!.remove(imageItem)
                activity!!.runOnUiThread {
                    val file = File(imageItem.userPath)
                    file.delete()
                    Toast.makeText(context, "Deleted!", Toast.LENGTH_SHORT).show()
                    mainFetch!!.removeListener(this)
                    mainFetch!!.remove(download.id)
                }
            }

            override fun onDownloadBlockUpdated(download: Download, downloadBlock: DownloadBlock, totalBlocks: Int) {}
            override fun onQueued(download: Download, waitingOnNetwork: Boolean) {}
            override fun onAdded(download: Download) {}
            override fun onCompleted(download: Download) {
                imageItem.isDownloading = false
                imageItem.size = Helper.humanReadableByteCount(download.total)
                model!!.downloading(imageItem)
                listView!!.smoothScrollToPosition(listViewAdapter!!.getPositionOfItem(imageItem))
                model!!.select(imageItem)
                activity!!.runOnUiThread { Toast.makeText(context, "Downloaded!", Toast.LENGTH_SHORT).show() }
                mainFetch!!.removeListener(this)
                mainFetch!!.remove(download.id)
            }

            override fun onProgress(download: Download, etaInMilliSeconds: Long, downloadedBytesPerSecond: Long) {
                if (request.id == download.id) {
                    imageItem.progress = download.progress
                    imageItem.size = Helper.humanReadableByteCount(download.downloaded) + " / " + Helper.humanReadableByteCount(download.total) + " - " + Helper.humanReadableByteCount(downloadedBytesPerSecond) + "/s"
                    model!!.downloading(imageItem)
                    val progress = download.progress
                    Log.d("Fetch", "Progress Completed :$progress")
                }
            }

            override fun onPaused(download: Download) {}
            override fun onResumed(download: Download) {}
            override fun onCancelled(download: Download) {
                model!!.remove(imageItem)
                activity!!.runOnUiThread {
                    val file = File(imageItem.userPath)
                    file.delete()
                    Toast.makeText(context, "Deleted!", Toast.LENGTH_SHORT).show()
                    mainFetch!!.removeListener(this)
                    mainFetch!!.remove(download.id)
                }
            }

            override fun onRemoved(download: Download) {}
            override fun onDeleted(download: Download) {}
        }
        mainFetch!!.addListener(fetchListener)
        mainFetch!!.enqueue(request, Func { download: Request -> Log.i("lala", "added " + download.id) }, Func { error: Error -> Log.i("lala", "error  " + error.name) })
    }
}